### PR TITLE
Add CodeParrot 🦜 codebase

### PR DESCRIPTION
**Pull request by lvwerra**
# Add CodeParrot 🦜 codebase

This PR adds the CodeParrot 🦜 codebase to `examples/research_projects/`. The folder `scripts/` includes files for the following steps:

- data preprocessing
- model initialization
- training with accelerate
- validation loss
- evaluate on HumanEval benchmark

In addition the README gives an overview and highlights the results. The requirements file fixes the dependencies.

cc @LysandreJik @thomwolf

### CODE REVIEW

File: examples/research_projects/codeparrot/README.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+- exact deduplication
+- filtering files with max line length > 1000
+- filtering files with mean line length > 100
+- fraction of alphanumeric characters < 0.25
+- containing the word "auto-generated" or similar in the first 5 lines
+
+The script to process the full dataset can be found in `scripts/preprocessing.py`. Executing the script on 16 CPUs takes roughly 3h and removes 70% of the original dataset. The cleaned [train](https://huggingface.co/datasets/lvwerra/codeparrot-clean-train) and [validation](https://huggingface.co/datasets/lvwerra/codeparrot-clean-valid) splits are also available on the Hub.
+
+## Training
+The models are randomly initialized and trained from scratch. The initialization script can be found at `scripts/initialize.py`. The main training script is built with 🤗 Accelerate to scale across a wide range of platforms and infrastructure scales.
```
Comment (lewtun) [inline examples/research_projects/codeparrot/README.md:36]: Ditto as above re how to actually use the `initialize.py` script. Ideally someone would be able to just follow the README and reproduce your results without having to read the source code in detail

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+from datasets import load_dataset
+from torch.utils.data import IterableDataset
+from torch.utils.data.dataloader import DataLoader
+from torch.utils.tensorboard import SummaryWriter
+
+import transformers
+import wandb
+from accelerate import Accelerator
+from huggingface_hub import Repository
+from transformers import AdamW, AutoTokenizer, GPT2LMHeadModel, get_scheduler, set_seed
```
Comment (lewtun) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:15]: Maybe we already discussed this, but why don't we use `AutoModelForCausalLM` here instead of `GPT2LMHeadModel`?

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+
+class ConstantLengthDataset(IterableDataset):
```
Comment (lewtun) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:18]: Would be nice to add a docstring and document the args in the `__init__` here

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            all_token_ids = []
+            for tokenized_input in tokenized_inputs:
+                all_token_ids.extend(tokenized_input + [self.concat_token_id])
+            for i in range(0, len(all_token_ids), self.seq_length):
+                input_ids = all_token_ids[i : i + self.seq_length]
+                if len(input_ids) == self.seq_length:
+                    yield torch.tensor(input_ids)
+
+
+def setup_logging(project_name):
```
Comment (lewtun) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:59]: Ditto as above re a docstring

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        perplexity = float("inf")
+    return loss.item(), perplexity.item()
+
+
+# Accelerator
+accelerator = Accelerator()
+acc_state = {str(k): str(v) for k, v in accelerator.state.__dict__.items()}
+
+# Hyperparameters (codeparrot-small configs are in comments)
+project_name = "lvwerra/codeparrot"
```
Comment (lewtun) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:139]: Can we make this a CLI argument instead of having it hard-coded? 

Perhaps @LysandreJik can correct me, but if the goal is to have a reproducible project then I think it would be good to make this as transparent as possible to the end-user.

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+dataset_name = "../codeparrot-clean"
+config = {
```
Comment (lewtun) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:141]: As above, can these hard coded values be promoted to CLI args as default values?

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    "lr_scheduler_type": "cosine",
+    "num_warmup_steps": 750,  # 2000
+    "gradient_accumulation_steps": 16,  # 1
+    "gradient_checkpointing": True,  # False
+    "max_train_steps": 50_000,  # 150_000
+    "max_eval_steps": -1,
+    "seq_length": 1024,
+    "seed": 1,
+    "save_checkpoint_steps": 50_000,
+}  # 15_000
```
Comment (lewtun) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:156]: Dead code?

File: examples/research_projects/codeparrot/scripts/human_eval.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+from datasets import load_dataset, load_metric
+from tqdm import tqdm
+
+import transformers
+from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline, set_seed
+
+
+transformers.logging.set_verbosity_error()
+os.environ["HF_ALLOW_CODE_EVAL"] = "1"
```
Comment (lewtun) [inline examples/research_projects/codeparrot/scripts/human_eval.py:14]: Perhaps add a comment here to explain why we set these env vars?

File: examples/research_projects/codeparrot/scripts/human_eval.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+def complete_code(pipe, prompt, num_completions=1, **gen_kwargs):
+    """Complete prompt with text generation pipeline and return num_completions."""
+    prompt = pipe.tokenizer.eos_token + prompt
+    code_gens = pipe(prompt, num_return_sequences=num_completions, **gen_kwargs)
+    return [first_block(code_gen["generated_text"][len(prompt) :]) for code_gen in code_gens]
+
+
+# Settings
+gen_kwargs = {
```
Comment (lewtun) [inline examples/research_projects/codeparrot/scripts/human_eval.py:31]: Should we promote these to CLI arguments?

File: examples/research_projects/codeparrot/README.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
+# CodeParrot 🦜
+<p align="center">
+    <img src="images/code-highlighting-streamlit.png" alt="drawing" width="350"/>
+</p>
+
+## What is this about?
```
Comment (lewtun) [inline examples/research_projects/codeparrot/README.md:6]: Maybe add your GitHub handle here that indicates you are the author / maintainer of this script?

**Review by lewtun — APPROVED** [review]
Thanks a lot for adding this to the examples! Overall it looks good, but I think it's lacking some details that could aid with reproducibility:

- For each step (preprocessing, training, evaluation etc), I think it would be useful to have a CLI with some default values. Currently, there are some hard-coded variables / configs that require the end-user to read the source code to understand what's going on.
- Add docstrings to all the classes and functions.

A nice template to look at is the [DistilBERT project](https://github.com/huggingface/transformers/tree/master/examples/research_projects/distillation). 

PS I'm not sure what the conventions are for new research projects, so feel free to ignore these comments if the current PR is OK for @LysandreJik

**Comment by lvwerra** [comment]
> * For each step (preprocessing, training, evaluation etc), I think it would be useful to have a CLI with some default values. Currently, there are some hard-coded variables / configs that require the end-user to read the source code to understand what's going on.

The reason I went without a CLI is that most scripts are quite lean and a CLI would decrease readability. E.g. for initialisation would significantly longer with a CLI. Due to the required compute scale I also expect this to be run by "advanced" users. This reminded me that the training script has to be executed from the accelerate CLI so I'll add a remark about this in any case.

What do you think @LysandreJik?

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 7 above)
Comment (lvwerra) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:156]: no, this is the codeparrot-small setting (see comment at beginning of config)

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 2 above)
Comment (lvwerra) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:15]: Artifact - I'll fix this.

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 7 above)
Comment (lewtun) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:156]: Ah, I see. If you plan to keep this comment-based approach, then perhaps the `15_000` needs to be moved within the dict?

File: examples/research_projects/codeparrot/README.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+## Preprocessing
+The raw dataset contains many duplications so the dataset was deduplicated and filtered using the heuristics proposed in the Codex [paper](https://arxiv.org/abs/2107.03374):
+
+- exact deduplication
+- filtering files with max line length > 1000
+- filtering files with mean line length > 100
+- fraction of alphanumeric characters < 0.25
+- containing the word "auto-generated" or similar in the first 5 lines
+
+The script to process the full dataset can be found in `scripts/preprocessing.py`. Executing the script on 16 CPUs takes roughly 3h and removes 70% of the original dataset. The cleaned [train](https://huggingface.co/datasets/lvwerra/codeparrot-clean-train) and [validation](https://huggingface.co/datasets/lvwerra/codeparrot-clean-valid) splits are also available on the Hub.
```
Comment (LysandreJik) [inline examples/research_projects/codeparrot/README.md:33]: Are you sure these aren't vCPUs rather than actual CPUs? 16 CPUs would be very impressive!
```suggestion
The script to process the full dataset can be found in `scripts/preprocessing.py`. Executing the script on 16 vCPUs takes roughly 3h and removes 70% of the original dataset. The cleaned [train](https://huggingface.co/datasets/lvwerra/codeparrot-clean-train) and [validation](https://huggingface.co/datasets/lvwerra/codeparrot-clean-valid) splits are also available on the Hub.
```

File: examples/research_projects/codeparrot/README.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+## Training
+The models are randomly initialized and trained from scratch. The initialization script can be found at `scripts/initialize.py`. The main training script is built with 🤗 Accelerate to scale across a wide range of platforms and infrastructure scales.
+
+We train two models with [110M](https://huggingface.co/lvwerra/codeparrot-small/) and [1.5B](https://huggingface.co/lvwerra/codeparrot/) parameters for 25-30B tokens on a 16xA100 (40GB) machine which takes 1 day and 1 week, respectively. The training script can be found in `scripts/codeparrot_training.py`.
+
+## Evaluation
+The validation loss can be calculate with the `scripts/validation_loss.py` script. In addition we evaluate the model on OpenAI's _HumanEval_ benchmark. The evaluation script can be found in `scripts/human_eval.py`. The results as well as reference values are shown in the following table:
```
Comment (LysandreJik) [inline examples/research_projects/codeparrot/README.md:41]:
```suggestion
The validation loss can be calculated with the `scripts/validation_loss.py` script. In addition, we evaluate the model on OpenAI's _HumanEval_ benchmark. The evaluation script can be found in `scripts/human_eval.py`. The results, as well as reference values, are shown in the following table:
```

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 5 above)
Comment (LysandreJik) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:139]: I agree :)

**Review by LysandreJik — COMMENTED** [review]
This looks great @lvwerra! Thanks for contributing your code, this is cool!

I agree with @lewtun that having the code as a CLI would be better for reusability, and I'd argue it would also be the best for readability: all arguments are centralized, making it easy to see what can be configured and what cannot.

I think people are also used to CLIs and it doesn't deter them from reading the code, so I personally see no drawbacks.

---

I see there is an image file in your PR. Could you please host it in a dataset on the hub, either on your profile or on the `hf-internal-testing` organization?

The issue with adding image files to the repository is that they'll weigh it down forever as they'll be part of its history.

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 5 above)
Comment (julien-c) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py:139]: FWIW we have a pretty cool class called `HfArgumentParser` to make CLI parsing easy to code (and to read)

File: examples/research_projects/codeparrot/README.md [hunk 13]
```diff
@@ -1,6 +1,6 @@
 # CodeParrot 🦜
 <p align="center">
-    <img src="images/code-highlighting-streamlit.png" alt="drawing" width="350"/>
+    <img src="https://huggingface.co/datasets/lvwerra/repo-images/raw/main/code-highlighting-streamlit.png" alt="drawing" width="350"/>
```
Comment (LysandreJik) [inline examples/research_projects/codeparrot/README.md:3]: Thank you!

File: examples/research_projects/codeparrot/scripts/bpe_training.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    for _ in tqdm(range(0, args.n_examples, batch_size)):
+        yield [next(iter_dataset)[args.text_column] for _ in range(batch_size)]
+
+
+# Configuration
+parser = HfArgumentParser(TokenizerTrainingArguments)
+args = parser.parse_args()
+
+# Base tokenizer
+tokenizer = AutoTokenizer.from_pretrained(args.base_tokenizer)
```
Comment (LysandreJik) [inline examples/research_projects/codeparrot/scripts/bpe_training.py:20]: Should this really be `AutoTokenizer` if you specify that it's BPE-only?

Comment (LysandreJik) [reply examples/research_projects/codeparrot/scripts/bpe_training.py:20]: (Not a big problem and I understand if you want to keep it that way)

**Review by LysandreJik — APPROVED** [review]
This looks great to me! Thanks for working on that, @lvwerra!

**Review by lewtun — APPROVED** [review]
Thanks for making the changes! I left a few minor nits, but otherwise LGTM!
